spring-cloud-kubernetes-client-autoconfig/src/main/java/org/springframework/cloud/kubernetes/client/KubernetesClientUtils.java

```diff
@@ -61,4 +65,42 @@ public static ApiClient kubernetesApiClient() {
 		}
 	}
 
+	/**
+	 * this method does the namespace resolution for both config map and secrets
+	 * implementations. It tries these places to find the namespace:
+	 *
+	 * <pre>
+	 *     1. from a normalized source (which can be null)
+	 *     2. from a property 'spring.cloud.kubernetes.client.namespace', if such is present
+	 *     3. from a String residing in a file denoted by `spring.cloud.kubernetes.client.serviceAccountNamespacePath`
+	 * 	      property, if such is present
+	 * 	   4. from a String residing in `/var/run/secrets/kubernetes.io/serviceaccount/namespace` file,
+	 * 	  	  if such is present (kubernetes default path)
+	 * </pre>
+	 *
+	 * If any of the above fail, we throw a NamespaceResolutionFailedException.
+	 * @param namespace normalized namespace
+	 * @param configurationTarget Config Map/Secret
+	 * @param provider the provider which computes the namespace
+	 * @return application namespace
+	 * @throws NamespaceResolutionFailedException when namespace could not be resolved
+	 */
+	public static String getApplicationNamespace(String namespace, String configurationTarget,
```

Comment:
this is the main change in this PR and it is needed in order to provide support for namespaces for the k8s-native catalog watch (that I will be working on). We did the same exact thing for the fabric8 client.